### PR TITLE
GHA release workflow, refactor PR and Daily workflows

### PR DESCRIPTION
**Pull request by bandish-shah**
# What does this PR do?

* Adds a `release.yaml` pipeline which can be triggered by the creation of a properly formatted tag (beginning with a 'v') or manually by repo admins.  The pipeline does the following:
  1. Runs code quality checks
  1. Builds source and wheel distributions
  1. Publishes PyPi production packages if triggered by by the `push` of a tag.  Otherwise builds and pushes packages to Test PyPi.
  1. Builds and pushes production Docker images if triggered by the `push` of a tag, skipped otherwise.  Production Docker build and push can be performed independently by triggering the `release-docker` pipeline manually.

* Adds a `release-docker.yaml` pipeline which builds and pushes production Docker images.  This flow can be triggered manually by repo admins

* Adds `coverage`, `pytest-cpu`, `pytest-gpu` and `docker-configure-build-push` reusable workflows

* Adds/refactors PR `pr-docker`, `pr-cpu`, `pr-gpu` pipelines to use reusable workflows

* Refacxtors `pytest-daily` to `daily` pipeline and uses reusable workflows

## Future Work (addressed by future PRs)

* Update `daily` to build Docker images and run tests on staged images
* Update Composer Docker builds to be gated by PyTorch images

# What issue(s) does this change relate to?

Closes [CO-1186](https://mosaicml.atlassian.net/browse/CO-1186)

# Testing

All workflows were tested manually by adding the `pull_request` trigger to all workflows and running in GHA.  This trigger has been removed from the `daily`, `release-docker` and `release` workflows.  The `pr-gpu` workflow has been changed to use the `pull_request_target` workflow since it needs to provide secrets to PR's from forked repos.  Links to passing GHA runs for these workflows can be found below.

`pr-gpu`: https://github.com/mosaicml/composer/actions/runs/4199439168

`daily`: https://github.com/mosaicml/composer/actions/runs/4208510415

`release-docker`: https://github.com/mosaicml/composer/actions/runs/4208510424

`release`: https://github.com/mosaicml/composer/actions/runs/4210173449

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-1186]: https://mosaicml.atlassian.net/browse/CO-1186?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/pr-docker.yaml
Comment (dakinggg) [inline .github/workflows/pr-docker.yaml:3]: I think this is frowned upon because it would allow someone to open a PR with a change to this file that dumps our secrets

File: .github/workflows/pr-gpu.yaml
Comment (dakinggg) [inline .github/workflows/pr-gpu.yaml:3]: same concern about secrets

File: .github/workflows/pr-gpu.yaml
Comment (dakinggg) [inline .github/workflows/pr-gpu.yaml:31]: Is this used anywhere?

File: .github/workflows/pr-cpu.yaml
Comment (dakinggg) [inline .github/workflows/pr-cpu.yaml:46]: Is this used anywhere?

File: .github/workflows/pytest-cpu.yaml
Comment (dakinggg) [inline .github/workflows/pytest-cpu.yaml:20]: Is this used anywhere?

File: .github/workflows/release.yaml
Comment (dakinggg) [inline .github/workflows/release.yaml:25]: maybe this should be 3.9 to be consistent with other places

File: .github/workflows/release-docker.yaml
Comment (dakinggg) [inline .github/workflows/release-docker.yaml:57]: I didn't quite understand. Isn't the "release" pipeline supposed to push docker images?

**Review by dakinggg — CHANGES_REQUESTED** [review]
Blocking because of my question about secrets, otherwise LGTM and we can ship and fix

File: .github/workflows/pr-docker.yaml
Comment (dakinggg) [reply .github/workflows/pr-docker.yaml:3]: Just to clarify, could someone change this workflow to run a (new) script which they check in and dumps secrets?

Comment (mvpatel2000) [reply .github/workflows/pr-docker.yaml:3]: If a PR changes this workflow, it won't run in that PR -- it has to be merged for it to run.

It's the same protection against someone adding a new workflow which uses `pull_request_target` that tries to dump secrets

Comment (dakinggg) [reply .github/workflows/pr-docker.yaml:3]: got it, thanks

Comment (dakinggg) [reply .github/workflows/pr-docker.yaml:3]: can we document what would be an unsafe thing to do directly in this yaml? I'd like to lower the likelihood of someone later changing this file and doing something unsafe

File: .github/workflows/release-docker.yaml
Comment (mvpatel2000) [reply .github/workflows/release-docker.yaml:57]: Also confused about this

File: .github/workflows/docker-configure-build-push.yaml
Comment (mvpatel2000) [inline .github/workflows/docker-configure-build-push.yaml:74]: Can we echo the full name we're using here a separate step called "[README] Docker Image Name" or something to that effect? Common use case is to open PR changing dockerfile, get staging image, and test, and it seems like it's currently a pain to actually dig through and find the right tag here

**Review by mvpatel2000 — COMMENTED** [review]
Mostly LGTM. Left 1 minor comment, and also confused by Daniel's points on python 3.9 and the push: false

File: .github/workflows/pr-docker.yaml
Comment (mvpatel2000) [reply .github/workflows/pr-docker.yaml:3]: Yea, we should definitely put a comment here on why we're using `pull_request_target` and why it's ok

Comment (bandish-shah) [reply .github/workflows/pr-docker.yaml:3]: Yea it's not ideal, there are two safeguards:
1. CI runs for PRs from external contributors must be approved by a maintainer
2. `pull_request_target` runs in the context of the target repo, not the merge commit 

1) safeguards against a malicious external user running anything (assuming the reviewer is diligent) and 2) prevents the external user from running malicious code (if the reviewer isn't diligent) unless that malicious code has already been committed to the target repo first.  It's not fool proof but it does make the case better.  More details on how all this works can be found here: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

We could also just assume that anyone making changes to workflows already has write access to the repo and change this to a `pull_request` trigger.  It would just mean that maintainers doing CI infra work can't work from forks which is probably ok.  Let me know what you think.

File: .github/workflows/pr-gpu.yaml
Comment (bandish-shah) [reply .github/workflows/pr-gpu.yaml:3]: This unfortunately does need `pull_request_target` since the MCloud API secret needs to be passed.  In addition to 1) and 2) previously, the other safeguard in this case is that malicious code is actually sandboxed within the CI cluster so the damage is contained.  We do not pass any GH permissions to the code running on the MCloud CI cluster so this is probably fine.

File: .github/workflows/pr-gpu.yaml
Comment (bandish-shah) [reply .github/workflows/pr-gpu.yaml:31]: Yea these are inputs to the `pytest-gpu.yaml` reusable workflow.  The syntax is a bit confusing but think about it as the `pr-gpu.yaml` instantiates the `pytest-gpu.yaml` workflow and these are input arguments.

File: .github/workflows/pr-cpu.yaml
Comment (bandish-shah) [reply .github/workflows/pr-cpu.yaml:46]: Inputs to `pytest-cpu.yaml` workflow.

File: .github/workflows/pytest-cpu.yaml
Comment (bandish-shah) [reply .github/workflows/pytest-cpu.yaml:20]: Good catch, it's not but it probably should be.  Fix coming.

File: .github/workflows/release-docker.yaml
Comment (bandish-shah) [reply .github/workflows/release-docker.yaml:57]: Good catch, had that as false before bc I didn't want to accidentally push new images.

File: .github/workflows/pr-docker.yaml
Comment (bandish-shah) [reply .github/workflows/pr-docker.yaml:3]: I'm actually just going to go ahead and change this to `pull_request`.

File: .github/workflows/pytest-cpu.yaml
Comment (mvpatel2000) [inline .github/workflows/pytest-cpu.yaml:48]: This should be removed so it doesn't override the python version in CPU image

**Review by mvpatel2000 — APPROVED** [review]
LGTM. There's enough moving pieces though that I would not merge until @dakinggg  also approves

**Review by dakinggg — APPROVED** [review]
LGTM, lets see if it works!

**Pull request merged**
